## - Creating <font color='blue'>m18_Global_QLQ30groups65gl</font> feature based on M18_Global_QLQ30
## - Using features which affect QLQ-C30 from <u>bibliography/research</u> ,  <font color='blue'>m0</font>, <font color='blue'>m3</font> and <font color='blue'>m6</font>
## - SVM - Support Vector Machine
## - SMOTE and without SMOTE 

In [100]:
#importing libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import pyreadstat

In [101]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

In [102]:
cols = ['age','cancer_stage','M0_performancestatus','M3_performancestatus','M6_performancestatus','income_3','M0_optimism_LOT','radiotherapy','endocrine','antiher2','chemotherapy','care_team','education_3','education_2','M0_employment_status','M3_employment_status','M6_employment_status','marital_status','marital_status_2','cancer_grade','M0_surgery','M3_surgery','M6_surgery','M3_emotional_support','M3_social_support_total','M0_employment_status','M6_employment_status','income_3','M0_mental_health_total_HADS','M3_mental_health_total_HADS','M6_mental_health_total_HADS','nccn_distress_thermometer','m3_nccn_distress_thermometer','M0_exercise','M6_exercise','M3_PTGI_spiritual_change','baseline_histological_type','baseline_chronic_illness','m0_bmi','m0_Global_QLQ30groups65gl','m3_Global_QLQ30groups65gl','m6_Global_QLQ30groups65gl','M18_Global_QLQ30']
df=pd.read_spss(r"C:\Users\Chris\Desktop\Cancer_Prediction\BOUNCE_Full Dataset_m0_m3_m6_m9_m12_m15_m18_09-12-2021_ORIGINAL.sav", usecols=cols)

In [103]:
df

age care_team  education_3                        education_2  \
0    66.0     CHAMP          1.0                          0-9 years   
1    54.0     CHAMP          2.0  high-school/vocational/University   
2    45.0     CHAMP          2.0  high-school/vocational/University   
3    70.0     CHAMP          1.0                          0-9 years   
4    49.0     CHAMP          3.0  high-school/vocational/University   
..    ...       ...          ...                                ...   
761   NaN       IEO          NaN                                NaN   
762  43.0       IEO          3.0  high-school/vocational/University   
763   NaN       IEO          NaN                                NaN   
764   NaN     CHAMP          NaN                                NaN   
765   NaN      HUJI          NaN                                NaN   

         marital_status                      marital_status_2  \
0               Widowed                Single/widowed/engaged   
1               Widowed                Single/widowed/engaged   
2    Common law partner  married/separated/common law partner   
3               Widowed                Single/widowed/engaged   
4               Married  married/separated/common law partner   
..                  ...                                   ...   
761                 NaN                                   NaN   
762  Common law partner  married/separated/common law partner   
763                 NaN                                   NaN   
764                 NaN                                   NaN   
765                 NaN                                   NaN   

    M0_employment_status M6_employment_status  income_3 M0_exercise  ...  \
0     employed full time   Employed full time  very low    moderate  ...   
1     employed full time   Employed full time   average          no  ...   
2          self-employed                  NaN  very low          no  ...   
3              housewife            Housewife  very low          no  ...   
4              housewife            Housewife      high    moderate  ...   
..                   ...                  ...       ...         ...  ...   
761                  NaN                  NaN       NaN         NaN  ...   
762        self-employed        Self employed   average    moderate  ...   
763                  NaN                  NaN       NaN         NaN  ...   
764                  NaN                  NaN       NaN         NaN  ...   
765                  NaN                  NaN       NaN         NaN  ...   

    chemotherapy  radiotherapy  endocrine  M0_surgery  M3_surgery  M6_surgery  \
0             No           Yes        Yes         1.0         0.0         0.0   
1            Yes           Yes        Yes         0.0         0.0         0.0   
2            Yes           Yes         No         0.0         NaN         NaN   
3             No           Yes        Yes         0.0         0.0         0.0   
4             No           Yes        Yes         0.0         0.0         0.0   
..           ...           ...        ...         ...         ...         ...   
761           No           Yes        Yes         NaN         0.0         NaN   
762          Yes            No        Yes         0.0         0.0         0.0   
763          Yes           Yes        Yes         0.0         0.0         0.0   
764          NaN           NaN        NaN         NaN         NaN         NaN   
765          NaN           NaN        NaN         NaN         NaN         NaN   

     m0_Global_QLQ30groups65gl  m3_Global_QLQ30groups65gl  \
0                     good QoL                   good QoL   
1                     good QoL                   good QoL   
2                     good QoL                        NaN   
3                     good QoL                   good QoL   
4                     good QoL                   good QoL   
..                         ...                        ...   
761                        NaN                   good QoL  

In [104]:
# create a list of our conditions (breakpoint 65)
# >65  ->  0
# <65  ->  1
conditions = [(df['M18_Global_QLQ30'] >= 0) & (df['M18_Global_QLQ30'] <= 65),(df['M18_Global_QLQ30'] > 65) & (df['M18_Global_QLQ30'] <= 100)]

In [105]:
# create a list of the values we want to assign for each condition
values = [9, 8]

In [106]:
df['m18_Global_QLQ30groups65gl'] = np.select(conditions, values)

In [107]:
df

age care_team  education_3                        education_2  \
0    66.0     CHAMP          1.0                          0-9 years   
1    54.0     CHAMP          2.0  high-school/vocational/University   
2    45.0     CHAMP          2.0  high-school/vocational/University   
3    70.0     CHAMP          1.0                          0-9 years   
4    49.0     CHAMP          3.0  high-school/vocational/University   
..    ...       ...          ...                                ...   
761   NaN       IEO          NaN                                NaN   
762  43.0       IEO          3.0  high-school/vocational/University   
763   NaN       IEO          NaN                                NaN   
764   NaN     CHAMP          NaN                                NaN   
765   NaN      HUJI          NaN                                NaN   

         marital_status                      marital_status_2  \
0               Widowed                Single/widowed/engaged   
1               Widowed                Single/widowed/engaged   
2    Common law partner  married/separated/common law partner   
3               Widowed                Single/widowed/engaged   
4               Married  married/separated/common law partner   
..                  ...                                   ...   
761                 NaN                                   NaN   
762  Common law partner  married/separated/common law partner   
763                 NaN                                   NaN   
764                 NaN                                   NaN   
765                 NaN                                   NaN   

    M0_employment_status M6_employment_status  income_3 M0_exercise  ...  \
0     employed full time   Employed full time  very low    moderate  ...   
1     employed full time   Employed full time   average          no  ...   
2          self-employed                  NaN  very low          no  ...   
3              housewife            Housewife  very low          no  ...   
4              housewife            Housewife      high    moderate  ...   
..                   ...                  ...       ...         ...  ...   
761                  NaN                  NaN       NaN         NaN  ...   
762        self-employed        Self employed   average    moderate  ...   
763                  NaN                  NaN       NaN         NaN  ...   
764                  NaN                  NaN       NaN         NaN  ...   
765                  NaN                  NaN       NaN         NaN  ...   

    radiotherapy  endocrine  M0_surgery  M3_surgery  M6_surgery  \
0            Yes        Yes         1.0         0.0         0.0   
1            Yes        Yes         0.0         0.0         0.0   
2            Yes         No         0.0         NaN         NaN   
3            Yes        Yes         0.0         0.0         0.0   
4            Yes        Yes         0.0         0.0         0.0   
..           ...        ...         ...         ...         ...   
761          Yes        Yes         NaN         0.0         NaN   
762           No        Yes         0.0         0.0         0.0   
763          Yes        Yes         0.0         0.0         0.0   
764          NaN        NaN         NaN         NaN         NaN   
765          NaN        NaN         NaN         NaN         NaN   

     m0_Global_QLQ30groups65gl  m3_Global_QLQ30groups65gl  \
0                     good QoL                   good QoL   
1                     good QoL                   good QoL   
2                     good QoL                        NaN   
3                     good QoL                   good QoL   
4                     good QoL                   good QoL   
..                         ...                        ...   
761                        NaN                   good QoL   
762                   good QoL                   poor QoL   
763                        NaN                   good QoL   
764                        NaN              

In [108]:
df['m18_Global_QLQ30groups65gl']=df['m18_Global_QLQ30groups65gl'].replace(0, np.NaN)
df['m18_Global_QLQ30groups65gl']=df['m18_Global_QLQ30groups65gl'].replace(9, 1)
df['m18_Global_QLQ30groups65gl']=df['m18_Global_QLQ30groups65gl'].replace(8, 0)

In [109]:
df = df.drop('M18_Global_QLQ30', 1)

C:\Users\Chris\AppData\Local\Temp/ipykernel_5256/2872487555.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  df = df.drop('M18_Global_QLQ30', 1)


In [110]:
df

age care_team  education_3                        education_2  \
0    66.0     CHAMP          1.0                          0-9 years   
1    54.0     CHAMP          2.0  high-school/vocational/University   
2    45.0     CHAMP          2.0  high-school/vocational/University   
3    70.0     CHAMP          1.0                          0-9 years   
4    49.0     CHAMP          3.0  high-school/vocational/University   
..    ...       ...          ...                                ...   
761   NaN       IEO          NaN                                NaN   
762  43.0       IEO          3.0  high-school/vocational/University   
763   NaN       IEO          NaN                                NaN   
764   NaN     CHAMP          NaN                                NaN   
765   NaN      HUJI          NaN                                NaN   

         marital_status                      marital_status_2  \
0               Widowed                Single/widowed/engaged   
1               Widowed                Single/widowed/engaged   
2    Common law partner  married/separated/common law partner   
3               Widowed                Single/widowed/engaged   
4               Married  married/separated/common law partner   
..                  ...                                   ...   
761                 NaN                                   NaN   
762  Common law partner  married/separated/common law partner   
763                 NaN                                   NaN   
764                 NaN                                   NaN   
765                 NaN                                   NaN   

    M0_employment_status M6_employment_status  income_3 M0_exercise  ...  \
0     employed full time   Employed full time  very low    moderate  ...   
1     employed full time   Employed full time   average          no  ...   
2          self-employed                  NaN  very low          no  ...   
3              housewife            Housewife  very low          no  ...   
4              housewife            Housewife      high    moderate  ...   
..                   ...                  ...       ...         ...  ...   
761                  NaN                  NaN       NaN         NaN  ...   
762        self-employed        Self employed   average    moderate  ...   
763                  NaN                  NaN       NaN         NaN  ...   
764                  NaN                  NaN       NaN         NaN  ...   
765                  NaN                  NaN       NaN         NaN  ...   

    chemotherapy  radiotherapy  endocrine  M0_surgery  M3_surgery  M6_surgery  \
0             No           Yes        Yes         1.0         0.0         0.0   
1            Yes           Yes        Yes         0.0         0.0         0.0   
2            Yes           Yes         No         0.0         NaN         NaN   
3             No           Yes        Yes         0.0         0.0         0.0   
4             No           Yes        Yes         0.0         0.0         0.0   
..           ...           ...        ...         ...         ...         ...   
761           No           Yes        Yes         NaN         0.0         NaN   
762          Yes            No        Yes         0.0         0.0         0.0   
763          Yes           Yes        Yes         0.0         0.0         0.0   
764          NaN           NaN        NaN         NaN         NaN         NaN   
765          NaN           NaN        NaN         NaN         NaN         NaN   

     m0_Global_QLQ30groups65gl  m3_Global_QLQ30groups65gl  \
0                     good QoL                   good QoL   
1                     good QoL                   good QoL   
2                     good QoL                        NaN   
3                     good QoL                   good QoL   
4                     good QoL                   good QoL   
..                         ...                        ...   
761                        NaN                   good QoL  

<b>0</b> -> <font color='blue'>good QoL</font> 
<br>
<b>1</b> -> <font color='red'>poor QoL</font> 

In [111]:
#check the category features
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 35 columns):
 #   Column                        Non-Null Count  Dtype   
---  ------                        --------------  -----   
 0   age                           757 non-null    float64 
 1   care_team                     766 non-null    category
 2   education_3                   750 non-null    float64 
 3   education_2                   750 non-null    category
 4   marital_status                745 non-null    category
 5   marital_status_2              745 non-null    category
 6   M0_employment_status          756 non-null    category
 7   M6_employment_status          626 non-null    category
 8   income_3                      710 non-null    category
 9   M0_exercise                   749 non-null    category
 10  M6_exercise                   616 non-null    category
 11  nccn_distress_thermometer     595 non-null    float64 
 12  m3_nccn_distress_thermometer  617 non-null    floa

In [112]:
#convert category to numeric value
from sklearn.preprocessing import LabelEncoder
labelencoder_Y = LabelEncoder()
df.iloc[:,1] = labelencoder_Y.fit_transform(df.iloc[:,1].values)
df.iloc[:,3] = labelencoder_Y.fit_transform(df.iloc[:,3].values)
df.iloc[:,4] = labelencoder_Y.fit_transform(df.iloc[:,4].values)
df.iloc[:,5] = labelencoder_Y.fit_transform(df.iloc[:,5].values)
df.iloc[:,6] = labelencoder_Y.fit_transform(df.iloc[:,6].values)
df.iloc[:,7] = labelencoder_Y.fit_transform(df.iloc[:,7].values)
df.iloc[:,8] = labelencoder_Y.fit_transform(df.iloc[:,8].values)
df.iloc[:,9] = labelencoder_Y.fit_transform(df.iloc[:,9].values)
df.iloc[:,10] = labelencoder_Y.fit_transform(df.iloc[:,10].values)
df.iloc[:,22] = labelencoder_Y.fit_transform(df.iloc[:,22].values)
df.iloc[:,23] = labelencoder_Y.fit_transform(df.iloc[:,23].values)
df.iloc[:,25] = labelencoder_Y.fit_transform(df.iloc[:,25].values)
df.iloc[:,26] = labelencoder_Y.fit_transform(df.iloc[:,26].values)
df.iloc[:,27] = labelencoder_Y.fit_transform(df.iloc[:,27].values)
df.iloc[:,31] = labelencoder_Y.fit_transform(df.iloc[:,31].values)
df.iloc[:,32] = labelencoder_Y.fit_transform(df.iloc[:,32].values)
df.iloc[:,33] = labelencoder_Y.fit_transform(df.iloc[:,33].values)

In [113]:
#check the category features
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 35 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   age                           757 non-null    float64
 1   care_team                     766 non-null    int32  
 2   education_3                   750 non-null    float64
 3   education_2                   766 non-null    int32  
 4   marital_status                766 non-null    int32  
 5   marital_status_2              766 non-null    int32  
 6   M0_employment_status          766 non-null    int32  
 7   M6_employment_status          766 non-null    int32  
 8   income_3                      766 non-null    int32  
 9   M0_exercise                   766 non-null    int32  
 10  M6_exercise                   766 non-null    int32  
 11  nccn_distress_thermometer     595 non-null    float64
 12  m3_nccn_distress_thermometer  617 non-null    float64
 13  M0_op

In [114]:
#convert corresponding int32 feature type to float64
df['care_team'] = df.care_team.astype(float)
df['education_2'] = df.education_2.astype(float)
df['marital_status'] = df.marital_status.astype(float)
df['marital_status_2'] = df.marital_status_2.astype(float)
df['M0_employment_status'] = df.M0_employment_status.astype(float)
df['M6_employment_status'] = df.M6_employment_status.astype(float)
df['income_3'] = df.income_3.astype(float)
df['M0_exercise'] = df.M0_exercise.astype(float)
df['M6_exercise'] = df.M6_exercise.astype(float)
df['baseline_chronic_illness'] = df.baseline_chronic_illness.astype(float)
df['baseline_histological_type'] = df.baseline_histological_type.astype(float)
df['chemotherapy'] = df.chemotherapy.astype(float)
df['radiotherapy'] = df.radiotherapy.astype(float)
df['endocrine'] = df.endocrine.astype(float)
df['m0_Global_QLQ30groups65gl'] = df.m0_Global_QLQ30groups65gl.astype(float)
df['m3_Global_QLQ30groups65gl'] = df.m3_Global_QLQ30groups65gl.astype(float)
df['m6_Global_QLQ30groups65gl'] = df.m6_Global_QLQ30groups65gl.astype(float)

In [115]:
#check the category features
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 35 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   age                           757 non-null    float64
 1   care_team                     766 non-null    float64
 2   education_3                   750 non-null    float64
 3   education_2                   766 non-null    float64
 4   marital_status                766 non-null    float64
 5   marital_status_2              766 non-null    float64
 6   M0_employment_status          766 non-null    float64
 7   M6_employment_status          766 non-null    float64
 8   income_3                      766 non-null    float64
 9   M0_exercise                   766 non-null    float64
 10  M6_exercise                   766 non-null    float64
 11  nccn_distress_thermometer     595 non-null    float64
 12  m3_nccn_distress_thermometer  617 non-null    float64
 13  M0_op

In [116]:
df

age  care_team  education_3  education_2  marital_status  \
0    66.0        0.0          1.0          0.0             4.0   
1    54.0        0.0          2.0          1.0             4.0   
2    45.0        0.0          2.0          1.0             0.0   
3    70.0        0.0          1.0          0.0             4.0   
4    49.0        0.0          3.0          1.0             1.0   
..    ...        ...          ...          ...             ...   
761   NaN        3.0          NaN          2.0             5.0   
762  43.0        3.0          3.0          1.0             0.0   
763   NaN        3.0          NaN          2.0             5.0   
764   NaN        0.0          NaN          2.0             5.0   
765   NaN        1.0          NaN          2.0             5.0   

     marital_status_2  M0_employment_status  M6_employment_status  income_3  \
0                 0.0                   0.0                   0.0       2.0   
1                 0.0                   0.0                   0.0       0.0   
2                 1.0                   4.0                   6.0       2.0   
3                 0.0                   2.0                   2.0       2.0   
4                 1.0                   2.0                   2.0       1.0   
..                ...                   ...                   ...       ...   
761               2.0                   6.0                   6.0       3.0   
762               1.0                   4.0                   4.0       0.0   
763               2.0                   6.0                   6.0       3.0   
764               2.0                   6.0                   6.0       3.0   
765               2.0                   6.0                   6.0       3.0   

     M0_exercise  ...  chemotherapy  radiotherapy  endocrine  M0_surgery  \
0            1.0  ...           0.0           1.0        1.0         1.0   
1            2.0  ...           1.0           1.0        1.0         0.0   
2            2.0  ...           1.0           1.0        0.0         0.0   
3            2.0  ...           0.0           1.0        1.0         0.0   
4            1.0  ...           0.0           1.0        1.0         0.0   
..           ...  ...           ...           ...        ...         ...   
761          3.0  ...           0.0           1.0        1.0         NaN   
762          1.0  ...           1.0           0.0        1.0         0.0   
763          3.0  ...           1.0           1.0        1.0         0.0   
764          3.0  ...           2.0           2.0        2.0         NaN   
765          3.0  ...           2.0           2.0        2.0         NaN   

     M3_surgery  M6_surgery  m0_Global_QLQ30groups65gl  \
0           0.0         0.0                        0.0   
1           0.0         0.0                        0.0   
2           NaN         NaN                        0.0   
3           0.0         0.0                        0.0   
4           0.0         0.0                        0.0   
..          ...         ...                        ...   
761         0.0         NaN                        2.0   
762         0.0         0.0                        0.0   
763         0.0         0.0                        2.0   
764         NaN         NaN                        2.0   
765         NaN         NaN                        2.0   

     m3_Global_QLQ30groups65gl  m6_Global_QLQ30groups65gl  \
0                          0.0                        0.0   
1                          0.0                        1.0   
2                          2.0                        2.0   
3                          0.0                        0.0   
4                          0.0                        0.0   
..                         ...                        ...   
761                        0.0                        2.0   
762                        1.0                        2.0   
763                        0.0                        0.0   
764                        2.0              

In [117]:
#convert corresponding numeric value to NaN
df['education_2']=df['education_2'].replace(2, np.NaN)
df['marital_status']=df['marital_status'].replace(5, np.NaN)
df['marital_status_2']=df['marital_status_2'].replace(2, np.NaN)
df['M0_employment_status']=df['M0_employment_status'].replace(6, np.NaN)
df['M6_employment_status']=df['M6_employment_status'].replace(6, np.NaN)
df['income_3']=df['income_3'].replace(3, np.NaN)
df['M0_exercise']=df['M0_exercise'].replace(3, np.NaN)
df['M6_exercise']=df['M6_exercise'].replace(3, np.NaN)
df['baseline_chronic_illness']=df['baseline_chronic_illness'].replace(2, np.NaN)
df['baseline_histological_type']=df['baseline_histological_type'].replace(3, np.NaN)
df['chemotherapy']=df['chemotherapy'].replace(2, np.NaN)
df['radiotherapy']=df['radiotherapy'].replace(2, np.NaN)
df['endocrine']=df['endocrine'].replace(2, np.NaN)
df['m0_Global_QLQ30groups65gl']=df['m0_Global_QLQ30groups65gl'].replace(2, np.NaN)
df['m3_Global_QLQ30groups65gl']=df['m3_Global_QLQ30groups65gl'].replace(2, np.NaN)
df['m6_Global_QLQ30groups65gl']=df['m6_Global_QLQ30groups65gl'].replace(2, np.NaN)

In [118]:
df

age  care_team  education_3  education_2  marital_status  \
0    66.0        0.0          1.0          0.0             4.0   
1    54.0        0.0          2.0          1.0             4.0   
2    45.0        0.0          2.0          1.0             0.0   
3    70.0        0.0          1.0          0.0             4.0   
4    49.0        0.0          3.0          1.0             1.0   
..    ...        ...          ...          ...             ...   
761   NaN        3.0          NaN          NaN             NaN   
762  43.0        3.0          3.0          1.0             0.0   
763   NaN        3.0          NaN          NaN             NaN   
764   NaN        0.0          NaN          NaN             NaN   
765   NaN        1.0          NaN          NaN             NaN   

     marital_status_2  M0_employment_status  M6_employment_status  income_3  \
0                 0.0                   0.0                   0.0       2.0   
1                 0.0                   0.0                   0.0       0.0   
2                 1.0                   4.0                   NaN       2.0   
3                 0.0                   2.0                   2.0       2.0   
4                 1.0                   2.0                   2.0       1.0   
..                ...                   ...                   ...       ...   
761               NaN                   NaN                   NaN       NaN   
762               1.0                   4.0                   4.0       0.0   
763               NaN                   NaN                   NaN       NaN   
764               NaN                   NaN                   NaN       NaN   
765               NaN                   NaN                   NaN       NaN   

     M0_exercise  ...  chemotherapy  radiotherapy  endocrine  M0_surgery  \
0            1.0  ...           0.0           1.0        1.0         1.0   
1            2.0  ...           1.0           1.0        1.0         0.0   
2            2.0  ...           1.0           1.0        0.0         0.0   
3            2.0  ...           0.0           1.0        1.0         0.0   
4            1.0  ...           0.0           1.0        1.0         0.0   
..           ...  ...           ...           ...        ...         ...   
761          NaN  ...           0.0           1.0        1.0         NaN   
762          1.0  ...           1.0           0.0        1.0         0.0   
763          NaN  ...           1.0           1.0        1.0         0.0   
764          NaN  ...           NaN           NaN        NaN         NaN   
765          NaN  ...           NaN           NaN        NaN         NaN   

     M3_surgery  M6_surgery  m0_Global_QLQ30groups65gl  \
0           0.0         0.0                        0.0   
1           0.0         0.0                        0.0   
2           NaN         NaN                        0.0   
3           0.0         0.0                        0.0   
4           0.0         0.0                        0.0   
..          ...         ...                        ...   
761         0.0         NaN                        NaN   
762         0.0         0.0                        0.0   
763         0.0         0.0                        NaN   
764         NaN         NaN                        NaN   
765         NaN         NaN                        NaN   

     m3_Global_QLQ30groups65gl  m6_Global_QLQ30groups65gl  \
0                          0.0                        0.0   
1                          0.0                        1.0   
2                          NaN                        NaN   
3                          0.0                        0.0   
4                          0.0                        0.0   
..                         ...                        ...   
761                        0.0                        NaN   
762                        1.0                        NaN   
763                        0.0                        0.0   
764                        NaN              

---------------

<font color='blue'><b>MICE Data Imoutation</b></font>

In [56]:
imputer = IterativeImputer(imputation_order='ascending',max_iter=10,random_state=42,n_nearest_features=None)

In [57]:
df = imputer.fit_transform(df)

c:\users\chris\appdata\local\programs\python\python39\lib\site-packages\sklearn\impute\_iterative.py:699: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.
  warnings.warn(


In [58]:
df

array([[66.        ,  0.        ,  1.        , ...,  0.        ,
         0.        ,  0.        ],
       [54.        ,  0.        ,  2.        , ...,  0.        ,
         1.        ,  0.        ],
       [45.        ,  0.        ,  2.        , ...,  0.32448779,
         0.22870169,  0.71528565],
       ...,
       [49.26714227,  3.        ,  2.37398529, ...,  0.        ,
         0.        ,  0.1993238 ],
       [53.54142224,  0.        ,  2.65636587, ...,  0.36959662,
         0.1230989 ,  0.        ],
       [55.21181968,  1.        ,  2.39826438, ...,  0.42606482,
         0.34475049,  1.        ]])

In [59]:
cols = len(df[0])
print('Number of columns', cols)

Number of columns 35


In [60]:
#Convert Array to Dataframe and name columns
df=pd.DataFrame(df,columns = ['age',
'care_team',
'education_3',
'education_2',
'marital_status',
'marital_status_2',
'M0_employment_status',
'M6_employment_status',
'income_3',
'M0_exercise',
'M6_exercise',
'nccn_distress_thermometer',
'm3_nccn_distress_thermometer',
'M0_optimism_LOT',
'M0_mental_health_total_HADS',
'M3_emotional_support',
'M3_social_support_total',
'M3_PTGI_spiritual_change',
'M3_mental_health_total_HADS',   
'M6_mental_health_total_HADS',
'cancer_stage',
'cancer_grade',
'baseline_chronic_illness',
'baseline_histological_type',
'M0_performancestatus',
'chemotherapy',
'radiotherapy',
'endocrine',
'M0_surgery',
'M3_surgery',
'M6_surgery',
'm0_Global_QLQ30groups65gl',
'm3_Global_QLQ30groups65gl',
'm6_Global_QLQ30groups65gl',
'm18_Global_QLQ30groups65gl'
])

In [61]:
df

age  care_team  education_3  education_2  marital_status  \
0    66.000000        0.0     1.000000     0.000000        4.000000   
1    54.000000        0.0     2.000000     1.000000        4.000000   
2    45.000000        0.0     2.000000     1.000000        0.000000   
3    70.000000        0.0     1.000000     0.000000        4.000000   
4    49.000000        0.0     3.000000     1.000000        1.000000   
..         ...        ...          ...          ...             ...   
761  53.933227        3.0     2.352726     0.895341        1.211033   
762  43.000000        3.0     3.000000     1.000000        0.000000   
763  49.267142        3.0     2.373985     0.897936        1.329394   
764  53.541422        0.0     2.656366     0.942647        1.513958   
765  55.211820        1.0     2.398264     0.917196        1.158782   

     marital_status_2  M0_employment_status  M6_employment_status  income_3  \
0            0.000000              0.000000              0.000000  2.000000   
1            0.000000              0.000000              0.000000  0.000000   
2            1.000000              4.000000              3.501426  2.000000   
3            0.000000              2.000000              2.000000  2.000000   
4            1.000000              2.000000              2.000000  1.000000   
..                ...                   ...                   ...       ...   
761          0.912354              1.353145              1.469328  0.690668   
762          1.000000              4.000000              4.000000  0.000000   
763          0.840229              0.995676              0.887505  0.531303   
764          0.816938              0.881806              1.073805  0.520119   
765          0.905314              1.567525              1.699787  0.805939   

     M0_exercise  ...  chemotherapy  radiotherapy  endocrine  M0_surgery  \
0       1.000000  ...      0.000000      1.000000   1.000000    1.000000   
1       2.000000  ...      1.000000      1.000000   1.000000    0.000000   
2       2.000000  ...      1.000000      1.000000   0.000000    0.000000   
3       2.000000  ...      0.000000      1.000000   1.000000    0.000000   
4       1.000000  ...      0.000000      1.000000   1.000000    0.000000   
..           ...  ...           ...           ...        ...         ...   
761     0.996128  ...      0.000000      1.000000   1.000000    0.292336   
762     1.000000  ...      1.000000      0.000000   1.000000    0.000000   
763     0.985937  ...      1.000000      1.000000   1.000000    0.000000   
764     0.932199  ...      0.631730      0.961305   0.848382    0.271470   
765     1.232448  ...      0.482578      0.725786   0.769720    0.191984   

     M3_surgery  M6_surgery  m0_Global_QLQ30groups65gl  \
0      0.000000    0.000000                   0.000000   
1      0.000000    0.000000                   0.000000   
2      0.012624    0.036539                   0.000000   
3      0.000000    0.000000                   0.000000   
4      0.000000    0.000000                   0.000000   
..          ...         ...                        ...   
761    0.000000   -0.008468                   0.162132   
762    0.000000    0.000000                   0.000000   
763    0.000000    0.000000                   0.355418   
764    0.007380    0.031380                   0.196583   
765    0.009526    0.025280                   0.358337   

     m3_Global_QLQ30groups65gl  m6_Global_QLQ30groups65gl  \
0                     0.000000                   0.000000   
1                     0.000000                   1.000000   
2                     0.324488                   0.228702   
3                     0.000000                   0.000000   
4                     0.000000                   0.000000   
..                         ...                        ...   
761                   0.000000                   0.024659   
762                   1.000000                   0.471597   
763                   0.000000                   0

<b>- roundup values</b>

In [62]:
df.age = df.age.round()
df.care_team = df.care_team.round()
df.education_3 = df.education_3.round()
df.education_2 = df.education_2.round()
df.marital_status = df.marital_status.round()
df.marital_status_2 = df.marital_status_2.round()
df.M0_employment_status = df.M0_employment_status.round()
df.M6_employment_status = df.M6_employment_status.round()
df.income_3 = df.income_3.round()
df.M0_exercise = df.M0_exercise.round()
df.M6_exercise = df.M6_exercise.round()
df.nccn_distress_thermometer = df.nccn_distress_thermometer.round()
df.m3_nccn_distress_thermometer = df.m3_nccn_distress_thermometer.round()
df.M0_optimism_LOT = df.M0_optimism_LOT.round()
df.M0_mental_health_total_HADS = df.M0_mental_health_total_HADS.round()
df.M3_emotional_support = df.M3_emotional_support.round()
df.M3_social_support_total = df.M3_social_support_total.round()
df.M3_PTGI_spiritual_change = df.M3_PTGI_spiritual_change.round()
df.M3_mental_health_total_HADS = df.M3_mental_health_total_HADS.round()
df.M6_mental_health_total_HADS = df.M6_mental_health_total_HADS.round()
df.cancer_stage = df.cancer_stage.round()
df.cancer_grade = df.cancer_grade.round()
df.baseline_chronic_illness = df.baseline_chronic_illness.round()
df.baseline_histological_type = df.baseline_histological_type.round()
df.M0_performancestatus = df.M0_performancestatus.round()
df.chemotherapy = df.chemotherapy.round()
df.radiotherapy = df.radiotherapy.round()
df.endocrine = df.endocrine.round()
df.M0_surgery = df.M0_surgery.round()
df.M3_surgery = df.M3_surgery.round()
df.M6_surgery = df.M6_surgery.round()
df.m0_Global_QLQ30groups65gl = df.m0_Global_QLQ30groups65gl.round()
df.m3_Global_QLQ30groups65gl = df.m3_Global_QLQ30groups65gl.round()
df.m6_Global_QLQ30groups65gl = df.m6_Global_QLQ30groups65gl.round()
df.m18_Global_QLQ30groups65gl = df.m18_Global_QLQ30groups65gl.round()

In [63]:
df

age  care_team  education_3  education_2  marital_status  \
0    66.0        0.0          1.0          0.0             4.0   
1    54.0        0.0          2.0          1.0             4.0   
2    45.0        0.0          2.0          1.0             0.0   
3    70.0        0.0          1.0          0.0             4.0   
4    49.0        0.0          3.0          1.0             1.0   
..    ...        ...          ...          ...             ...   
761  54.0        3.0          2.0          1.0             1.0   
762  43.0        3.0          3.0          1.0             0.0   
763  49.0        3.0          2.0          1.0             1.0   
764  54.0        0.0          3.0          1.0             2.0   
765  55.0        1.0          2.0          1.0             1.0   

     marital_status_2  M0_employment_status  M6_employment_status  income_3  \
0                 0.0                   0.0                   0.0       2.0   
1                 0.0                   0.0                   0.0       0.0   
2                 1.0                   4.0                   4.0       2.0   
3                 0.0                   2.0                   2.0       2.0   
4                 1.0                   2.0                   2.0       1.0   
..                ...                   ...                   ...       ...   
761               1.0                   1.0                   1.0       1.0   
762               1.0                   4.0                   4.0       0.0   
763               1.0                   1.0                   1.0       1.0   
764               1.0                   1.0                   1.0       1.0   
765               1.0                   2.0                   2.0       1.0   

     M0_exercise  ...  chemotherapy  radiotherapy  endocrine  M0_surgery  \
0            1.0  ...           0.0           1.0        1.0         1.0   
1            2.0  ...           1.0           1.0        1.0         0.0   
2            2.0  ...           1.0           1.0        0.0         0.0   
3            2.0  ...           0.0           1.0        1.0         0.0   
4            1.0  ...           0.0           1.0        1.0         0.0   
..           ...  ...           ...           ...        ...         ...   
761          1.0  ...           0.0           1.0        1.0         0.0   
762          1.0  ...           1.0           0.0        1.0         0.0   
763          1.0  ...           1.0           1.0        1.0         0.0   
764          1.0  ...           1.0           1.0        1.0         0.0   
765          1.0  ...           0.0           1.0        1.0         0.0   

     M3_surgery  M6_surgery  m0_Global_QLQ30groups65gl  \
0           0.0         0.0                        0.0   
1           0.0         0.0                        0.0   
2           0.0         0.0                        0.0   
3           0.0         0.0                        0.0   
4           0.0         0.0                        0.0   
..          ...         ...                        ...   
761         0.0        -0.0                        0.0   
762         0.0         0.0                        0.0   
763         0.0         0.0                        0.0   
764         0.0         0.0                        0.0   
765         0.0         0.0                        0.0   

     m3_Global_QLQ30groups65gl  m6_Global_QLQ30groups65gl  \
0                          0.0                        0.0   
1                          0.0                        1.0   
2                          0.0                        0.0   
3                          0.0                        0.0   
4                          0.0                        0.0   
..                         ...                        ...   
761                        0.0                        0.0   
762                        1.0                        0.0   
763                        0.0                        0.0   
764                        0.0              

-----------------

<font color='blue'><b><u>SVM - Support Vector Machine (ten fold)</u></b></font>

In [64]:
from sklearn.svm import SVC
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict

In [65]:
# split the dataset into independent (X) and dependent (Y) (m12) which is the target value
X=df.iloc[:,1:-1].values
Y=df.iloc[:,-1].values

In [66]:
X_train, X_test, Y_train, Y_test = X[train_index], X[test_index], Y[train_index], Y[test_index]

In [67]:
classifier = SVC(kernel='rbf')

In [68]:
from sklearn.metrics import accuracy_score
classifier.fit(X_train,Y_train)
Y_predict=classifier.predict(X_test)
print('Accuracy:',accuracy_score(Y_test,Y_predict))

Accuracy: 0.7532467532467533


<font color='red'>Accuracy: 75.32%</font>

In [69]:
y_predict = classifier.predict(X_test)

In [70]:
#confusion matrix
pd.crosstab(Y_test,y_predict)

col_0  0.0  1.0
row_0          
0.0     42    5
1.0     14   16

For 0: <font color='blue'>89%</font> (42/47) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>53%</font> (16/30) correctly classified -> <font color='red'>poor QoL</font> accuracy

----------------------

<u><font color='blue'>SMOTE</u></font>

In [71]:
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
smote = SMOTE()

In [72]:
X_train_smote, Y_train_smote = smote.fit_resample(X_train.astype('float'),Y_train)

In [73]:
from collections import Counter
print("Before SMOTE: ", Counter(Y_train))
print("After SMOTE: ", Counter(Y_train_smote))

Before SMOTE:  Counter({0.0: 436, 1.0: 253})
After SMOTE:  Counter({0.0: 436, 1.0: 436})


In [74]:
from sklearn.metrics import accuracy_score
classifier.fit(X_train_smote,Y_train_smote)
Y_predict=classifier.predict(X_test)
print('Accuracy:',accuracy_score(Y_test,Y_predict))

Accuracy: 0.7272727272727273


<font color='red'>Accuracy: 72,72%</font>

In [75]:
pd.crosstab(Y_test,Y_predict)

col_0  0.0  1.0
row_0          
0.0     37   10
1.0     11   19

For 0: <font color='blue'>79%</font> (37/47) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>63%</font> (19/30) correctly classified -> <font color='red'>poor QoL</font> accuracy

--------------------

<font color='blue'><b><u>without</b></u></font> MICE Data Imoutation <br>
replacing remaining NaN values with the corresponding <u>mean value</u>

In [119]:
df

age  care_team  education_3  education_2  marital_status  \
0    66.0        0.0          1.0          0.0             4.0   
1    54.0        0.0          2.0          1.0             4.0   
2    45.0        0.0          2.0          1.0             0.0   
3    70.0        0.0          1.0          0.0             4.0   
4    49.0        0.0          3.0          1.0             1.0   
..    ...        ...          ...          ...             ...   
761   NaN        3.0          NaN          NaN             NaN   
762  43.0        3.0          3.0          1.0             0.0   
763   NaN        3.0          NaN          NaN             NaN   
764   NaN        0.0          NaN          NaN             NaN   
765   NaN        1.0          NaN          NaN             NaN   

     marital_status_2  M0_employment_status  M6_employment_status  income_3  \
0                 0.0                   0.0                   0.0       2.0   
1                 0.0                   0.0                   0.0       0.0   
2                 1.0                   4.0                   NaN       2.0   
3                 0.0                   2.0                   2.0       2.0   
4                 1.0                   2.0                   2.0       1.0   
..                ...                   ...                   ...       ...   
761               NaN                   NaN                   NaN       NaN   
762               1.0                   4.0                   4.0       0.0   
763               NaN                   NaN                   NaN       NaN   
764               NaN                   NaN                   NaN       NaN   
765               NaN                   NaN                   NaN       NaN   

     M0_exercise  ...  chemotherapy  radiotherapy  endocrine  M0_surgery  \
0            1.0  ...           0.0           1.0        1.0         1.0   
1            2.0  ...           1.0           1.0        1.0         0.0   
2            2.0  ...           1.0           1.0        0.0         0.0   
3            2.0  ...           0.0           1.0        1.0         0.0   
4            1.0  ...           0.0           1.0        1.0         0.0   
..           ...  ...           ...           ...        ...         ...   
761          NaN  ...           0.0           1.0        1.0         NaN   
762          1.0  ...           1.0           0.0        1.0         0.0   
763          NaN  ...           1.0           1.0        1.0         0.0   
764          NaN  ...           NaN           NaN        NaN         NaN   
765          NaN  ...           NaN           NaN        NaN         NaN   

     M3_surgery  M6_surgery  m0_Global_QLQ30groups65gl  \
0           0.0         0.0                        0.0   
1           0.0         0.0                        0.0   
2           NaN         NaN                        0.0   
3           0.0         0.0                        0.0   
4           0.0         0.0                        0.0   
..          ...         ...                        ...   
761         0.0         NaN                        NaN   
762         0.0         0.0                        0.0   
763         0.0         0.0                        NaN   
764         NaN         NaN                        NaN   
765         NaN         NaN                        NaN   

     m3_Global_QLQ30groups65gl  m6_Global_QLQ30groups65gl  \
0                          0.0                        0.0   
1                          0.0                        1.0   
2                          NaN                        NaN   
3                          0.0                        0.0   
4                          0.0                        0.0   
..                         ...                        ...   
761                        0.0                        NaN   
762                        1.0                        NaN   
763                        0.0                        0.0   
764                        NaN              

In [120]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 35 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   age                           757 non-null    float64
 1   care_team                     766 non-null    float64
 2   education_3                   750 non-null    float64
 3   education_2                   750 non-null    float64
 4   marital_status                745 non-null    float64
 5   marital_status_2              745 non-null    float64
 6   M0_employment_status          756 non-null    float64
 7   M6_employment_status          626 non-null    float64
 8   income_3                      710 non-null    float64
 9   M0_exercise                   749 non-null    float64
 10  M6_exercise                   616 non-null    float64
 11  nccn_distress_thermometer     595 non-null    float64
 12  m3_nccn_distress_thermometer  617 non-null    float64
 13  M0_op

<b>- replace NaN values with mean of each column -</b>

In [121]:
df_mean = df.fillna(df.mean())

In [122]:
df_mean

age  care_team  education_3  education_2  marital_status  \
0    66.000000        0.0     1.000000     0.000000        4.000000   
1    54.000000        0.0     2.000000     1.000000        4.000000   
2    45.000000        0.0     2.000000     1.000000        0.000000   
3    70.000000        0.0     1.000000     0.000000        4.000000   
4    49.000000        0.0     3.000000     1.000000        1.000000   
..         ...        ...          ...          ...             ...   
761  54.108322        3.0     2.462667     0.921333        1.328859   
762  43.000000        3.0     3.000000     1.000000        0.000000   
763  54.108322        3.0     2.462667     0.921333        1.328859   
764  54.108322        0.0     2.462667     0.921333        1.328859   
765  54.108322        1.0     2.462667     0.921333        1.328859   

     marital_status_2  M0_employment_status  M6_employment_status  income_3  \
0             0.00000               0.00000              0.000000  2.000000   
1             0.00000               0.00000              0.000000  0.000000   
2             1.00000               4.00000              1.420128  2.000000   
3             0.00000               2.00000              2.000000  2.000000   
4             1.00000               2.00000              2.000000  1.000000   
..                ...                   ...                   ...       ...   
761           0.85906               1.32672              1.420128  0.630986   
762           1.00000               4.00000              4.000000  0.000000   
763           0.85906               1.32672              1.420128  0.630986   
764           0.85906               1.32672              1.420128  0.630986   
765           0.85906               1.32672              1.420128  0.630986   

     M0_exercise  ...  chemotherapy  radiotherapy  endocrine  M0_surgery  \
0       1.000000  ...      0.000000       1.00000   1.000000    1.000000   
1       2.000000  ...      1.000000       1.00000   1.000000    0.000000   
2       2.000000  ...      1.000000       1.00000   0.000000    0.000000   
3       2.000000  ...      0.000000       1.00000   1.000000    0.000000   
4       1.000000  ...      0.000000       1.00000   1.000000    0.000000   
..           ...  ...           ...           ...        ...         ...   
761     1.034713  ...      0.000000       1.00000   1.000000    0.264228   
762     1.000000  ...      1.000000       0.00000   1.000000    0.000000   
763     1.034713  ...      1.000000       1.00000   1.000000    0.000000   
764     1.034713  ...      0.501992       0.74574   0.816514    0.264228   
765     1.034713  ...      0.501992       0.74574   0.816514    0.264228   

     M3_surgery  M6_surgery  m0_Global_QLQ30groups65gl  \
0       0.00000    0.000000                   0.000000   
1       0.00000    0.000000                   0.000000   
2       0.00768    0.021138                   0.000000   
3       0.00000    0.000000                   0.000000   
4       0.00000    0.000000                   0.000000   
..          ...         ...                        ...   
761     0.00000    0.021138                   0.282967   
762     0.00000    0.000000                   0.000000   
763     0.00000    0.000000                   0.282967   
764     0.00768    0.021138                   0.282967   
765     0.00768    0.021138                   0.282967   

     m3_Global_QLQ30groups65gl  m6_Global_QLQ30groups65gl  \
0                     0.000000                   0.000000   
1                     0.000000                   1.000000   
2                     0.348765                   0.222222   
3                     0.000000                   0.000000   
4                     0.000000                   0.000000   
..                         ...                        ...   
761                   0.000000                   0.222222   
762                   1.000000                   0.222222   
763                   0.000000                   0

<b>- roundup values</b>  -

In [123]:
df_mean.age = df_mean.age.round()
df_mean.care_team = df_mean.care_team.round()
df_mean.education_3 = df_mean.education_3.round()
df_mean.education_2 = df_mean.education_2.round()
df_mean.marital_status = df_mean.marital_status.round()
df_mean.marital_status_2 = df_mean.marital_status_2.round()
df_mean.M0_employment_status = df_mean.M0_employment_status.round()
df_mean.M6_employment_status = df_mean.M6_employment_status.round()
df_mean.income_3 = df_mean.income_3.round()
df_mean.M0_exercise = df_mean.M0_exercise.round()
df_mean.M6_exercise = df_mean.M6_exercise.round()
df_mean.nccn_distress_thermometer = df_mean.nccn_distress_thermometer.round()
df_mean.m3_nccn_distress_thermometer = df_mean.m3_nccn_distress_thermometer.round()
df_mean.M0_optimism_LOT = df_mean.M0_optimism_LOT.round()
df_mean.M0_mental_health_total_HADS = df_mean.M0_mental_health_total_HADS.round()
df_mean.M3_emotional_support = df_mean.M3_emotional_support.round()
df_mean.M3_social_support_total = df_mean.M3_social_support_total.round()
df_mean.M3_PTGI_spiritual_change = df_mean.M3_PTGI_spiritual_change.round()
df_mean.M3_mental_health_total_HADS = df_mean.M3_mental_health_total_HADS.round()
df_mean.M6_mental_health_total_HADS = df_mean.M6_mental_health_total_HADS.round()
df_mean.cancer_stage = df_mean.cancer_stage.round()
df_mean.cancer_grade = df_mean.cancer_grade.round()
df_mean.baseline_chronic_illness = df_mean.baseline_chronic_illness.round()
df_mean.baseline_histological_type = df_mean.baseline_histological_type.round()
df_mean.M0_performancestatus = df_mean.M0_performancestatus.round()
df_mean.chemotherapy = df_mean.chemotherapy.round()
df_mean.radiotherapy = df_mean.radiotherapy.round()
df_mean.endocrine = df_mean.endocrine.round()
df_mean.M0_surgery = df_mean.M0_surgery.round()
df_mean.M3_surgery = df_mean.M3_surgery.round()
df_mean.M6_surgery = df_mean.M6_surgery.round()
df_mean.m0_Global_QLQ30groups65gl = df_mean.m0_Global_QLQ30groups65gl.round()
df_mean.m3_Global_QLQ30groups65gl = df_mean.m3_Global_QLQ30groups65gl.round()
df_mean.m6_Global_QLQ30groups65gl = df_mean.m6_Global_QLQ30groups65gl.round()
df_mean.m18_Global_QLQ30groups65gl = df_mean.m18_Global_QLQ30groups65gl.round()

In [124]:
df_mean

age  care_team  education_3  education_2  marital_status  \
0    66.0        0.0          1.0          0.0             4.0   
1    54.0        0.0          2.0          1.0             4.0   
2    45.0        0.0          2.0          1.0             0.0   
3    70.0        0.0          1.0          0.0             4.0   
4    49.0        0.0          3.0          1.0             1.0   
..    ...        ...          ...          ...             ...   
761  54.0        3.0          2.0          1.0             1.0   
762  43.0        3.0          3.0          1.0             0.0   
763  54.0        3.0          2.0          1.0             1.0   
764  54.0        0.0          2.0          1.0             1.0   
765  54.0        1.0          2.0          1.0             1.0   

     marital_status_2  M0_employment_status  M6_employment_status  income_3  \
0                 0.0                   0.0                   0.0       2.0   
1                 0.0                   0.0                   0.0       0.0   
2                 1.0                   4.0                   1.0       2.0   
3                 0.0                   2.0                   2.0       2.0   
4                 1.0                   2.0                   2.0       1.0   
..                ...                   ...                   ...       ...   
761               1.0                   1.0                   1.0       1.0   
762               1.0                   4.0                   4.0       0.0   
763               1.0                   1.0                   1.0       1.0   
764               1.0                   1.0                   1.0       1.0   
765               1.0                   1.0                   1.0       1.0   

     M0_exercise  ...  chemotherapy  radiotherapy  endocrine  M0_surgery  \
0            1.0  ...           0.0           1.0        1.0         1.0   
1            2.0  ...           1.0           1.0        1.0         0.0   
2            2.0  ...           1.0           1.0        0.0         0.0   
3            2.0  ...           0.0           1.0        1.0         0.0   
4            1.0  ...           0.0           1.0        1.0         0.0   
..           ...  ...           ...           ...        ...         ...   
761          1.0  ...           0.0           1.0        1.0         0.0   
762          1.0  ...           1.0           0.0        1.0         0.0   
763          1.0  ...           1.0           1.0        1.0         0.0   
764          1.0  ...           1.0           1.0        1.0         0.0   
765          1.0  ...           1.0           1.0        1.0         0.0   

     M3_surgery  M6_surgery  m0_Global_QLQ30groups65gl  \
0           0.0         0.0                        0.0   
1           0.0         0.0                        0.0   
2           0.0         0.0                        0.0   
3           0.0         0.0                        0.0   
4           0.0         0.0                        0.0   
..          ...         ...                        ...   
761         0.0         0.0                        0.0   
762         0.0         0.0                        0.0   
763         0.0         0.0                        0.0   
764         0.0         0.0                        0.0   
765         0.0         0.0                        0.0   

     m3_Global_QLQ30groups65gl  m6_Global_QLQ30groups65gl  \
0                          0.0                        0.0   
1                          0.0                        1.0   
2                          0.0                        0.0   
3                          0.0                        0.0   
4                          0.0                        0.0   
..                         ...                        ...   
761                        0.0                        0.0   
762                        1.0                        0.0   
763                        0.0                        0.0   
764                        0.0              

In [125]:
df_mean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 35 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   age                           766 non-null    float64
 1   care_team                     766 non-null    float64
 2   education_3                   766 non-null    float64
 3   education_2                   766 non-null    float64
 4   marital_status                766 non-null    float64
 5   marital_status_2              766 non-null    float64
 6   M0_employment_status          766 non-null    float64
 7   M6_employment_status          766 non-null    float64
 8   income_3                      766 non-null    float64
 9   M0_exercise                   766 non-null    float64
 10  M6_exercise                   766 non-null    float64
 11  nccn_distress_thermometer     766 non-null    float64
 12  m3_nccn_distress_thermometer  766 non-null    float64
 13  M0_op

-----------------

<font color='blue'><b><u>SVM - Support Vector Machine (ten fold)</u></b></font>

In [135]:
from sklearn.svm import SVC
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict

In [136]:
# split the dataset into independent (X) and dependent (Y) (m12) which is the target value
X=df_mean.iloc[:,1:-1].values
Y=df_mean.iloc[:,-1].values

In [137]:
X_train, X_test, Y_train, Y_test = X[train_index], X[test_index], Y[train_index], Y[test_index]

In [138]:
classifier = SVC(kernel='rbf')

In [139]:
from sklearn.metrics import accuracy_score
classifier.fit(X_train,Y_train)
Y_predict=classifier.predict(X_test)
print('Accuracy:',accuracy_score(Y_test,Y_predict))

Accuracy: 0.7142857142857143


<font color='red'>Accuracy: 71.43%</font>

In [140]:
y_predict = classifier.predict(X_test)

In [141]:
#confusion matrix
pd.crosstab(Y_test,y_predict)

col_0  0.0  1.0
row_0          
0.0     52    6
1.0     16    3

For 0: <font color='blue'>90%</font> (52/58) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>16%</font> (3/19) correctly classified -> <font color='red'>poor QoL</font> accuracy

-------

<u><font color='blue'>SMOTE</u></font>

In [142]:
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
smote = SMOTE()

In [143]:
X_train_smote, Y_train_smote = smote.fit_resample(X_train.astype('float'),Y_train)

In [144]:
from collections import Counter
print("Before SMOTE: ", Counter(Y_train))
print("After SMOTE: ", Counter(Y_train_smote))

Before SMOTE:  Counter({0.0: 514, 1.0: 175})
After SMOTE:  Counter({0.0: 514, 1.0: 514})


In [145]:
from sklearn.metrics import accuracy_score
classifier.fit(X_train_smote,Y_train_smote)
Y_predict=classifier.predict(X_test)
print('Accuracy:',accuracy_score(Y_test,Y_predict))

Accuracy: 0.6363636363636364


<font color='red'>Accuracy: 63,63%</font>

In [146]:
pd.crosstab(Y_test,Y_predict)

col_0  0.0  1.0
row_0          
0.0     40   18
1.0     10    9

For 0: <font color='blue'>69%</font> (40/58) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>47%</font> (9/19) correctly classified -> <font color='red'>poor QoL</font> accuracy